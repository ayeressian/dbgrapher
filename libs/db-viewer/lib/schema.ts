import type Point from "./point";

export interface FkSchema {
  table: string;
  column: string;
}
export interface ColumnNoneFkSchema {
  name: string;
  type: string;
  pk?: boolean;
  uq?: boolean;
  nn?: boolean;
}

export declare interface ColumnFkSchema
  extends Omit<ColumnNoneFkSchema, "type"> {
  fk: FkSchema;
}

export type ColumnSchema = ColumnFkSchema | ColumnNoneFkSchema;

export interface TableSchema {
  name: string;
  pos: Point;
  columns: ColumnSchema[];
}

export interface AnnotationSchema {
  pos: Point;
  title?: string;
  text?: string;
  color?: string;
  width: number;
  height: number;
}

export interface Schema {
  tables: TableSchema[];
  annotations?: AnnotationSchema[];
  arrangement?: TableArrang;
  viewport?: Viewport;
  viewWidth?: number;
  viewHeight?: number;
}

export const enum Viewport {
  noChange = "noChange",
  centerByTablesWeight = "centerByTablesWeight",
  center = "center",
  centerByTables = "centerByTables",
}

export enum TableArrang {
  spiral = "spiral",
  default = "default",
}

export const isColumnFk = (column: ColumnSchema): column is ColumnFkSchema => {
  return (column as ColumnFkSchema).fk !== undefined;
};
