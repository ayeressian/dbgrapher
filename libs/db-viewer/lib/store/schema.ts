import { get, writable, type Readable } from "svelte/store";
import type { Schema } from "../schema";
import type { Store } from "./store";
import type { Viewport } from "./view";

export type RectSize = {
  width: number;
  height: number;
};

export class SchemaStore {
  private store: Store;
  schema = writable<Schema>({ tables: [] });
  // allTableSizesSetStore: Readable<boolean>;

  constructor(store: Store) {
    this.store = store;
  }

  getSchema(): Schema {
    const schema = get(this.schema);
    this.store.table.updateTablePoses(schema.tables);
    return schema;
  }

  setSchema(schema: Schema, viewport: Viewport = "center") {
    schema = JSON.parse(JSON.stringify(schema));
    this.store?.view.viewportStore.set(viewport);
    const tableStore = this.store.table;
    tableStore.tableSizes.clear();
    tableStore.tablePoses.clear();
    tableStore.tableToHighlight.clear();
    tableStore.tableFromHighlight.clear();
    schema.tables.forEach((table) => {
      const tableSizeWritable = writable<RectSize>();
      tableStore.tableSizes.set(table.name, tableSizeWritable);
      const tablePosWritable = writable(table.pos);
      tableStore.tablePoses.set(table.name, tablePosWritable);
      tablePosWritable.subscribe(this.store!.view.getUpdateViewBound());
      tableStore.tableToHighlight.set(table.name, writable(""));
      tableStore.tableFromHighlight.set(table.name, writable(""));
    });
    this.schema.set({tables: []});
    setTimeout(() => this.schema.set(schema));    

    // allTableSizesSetStore = derived(
    //   [...tableSizes.values()],
    //   (tableSizes) => {
    //     return tableSizes.every((tableSize) => {
    //       return tableSize.height > 0 && tableSize.width > 0;
    //     });
    //   }
    // );
  }
}
