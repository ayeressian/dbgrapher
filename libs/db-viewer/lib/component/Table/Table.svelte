<script lang="ts">
  import type Point from "../../point";

  import Column, { Highlight } from "../Column/Column.svelte";
  import { getContext, onMount, tick } from "svelte";
  import type { ColumnFkSchema } from "../../schema";
  import type { Store } from "../../store/store";
  import type TableData from "../DBViewer/event";

  export let name: string;

  type EventCallback = (tableData: TableData) => void;

  export let mouseDown: EventCallback;
  export let click: EventCallback;
  export let dblClick: EventCallback;
  export let contextMenu: EventCallback;

  const store = getContext<Store>("store");
  const schemaStore = store.schema.schema;
  const zoomStore = store.view.zoom;

  let width = 0;
  let height = 0;

  let tableElem: HTMLTableElement;

  const table = $schemaStore.tables.find((table) => table.name === name);
  const tableSize = store.table.getTableSize(name)!;
  const tablePos = store.table.getTablePos(name)!;

  let initMouseDown: Point;
  let initPos: Point;

  let moveInProgress = false;

  const getTableData = () => ({ name, pos: $tablePos, size: $tableSize });

  onMount(async () => {
    const ro = new ResizeObserver(() => {
      width = tableElem.offsetWidth;
      height = tableElem.offsetHeight;
      tableSize.set({
        width,
        height,
      });
    });
    ro.observe(tableElem);
  });

  function mouseUp(): void {
    document.removeEventListener("mousemove", onMouseMove);
    moveInProgress = false;
  }

  function mouseDownInternal(event: MouseEvent): void {
    initMouseDown = {
      x: event.clientX * $zoomStore,
      y: event.clientY * $zoomStore,
    };

    initPos = $tablePos;

    moveInProgress = true;
    mouseDown(getTableData());
    document.addEventListener("mousemove", onMouseMove);

    document.addEventListener("mouseup", mouseUp);
  }

  function onMouseMove(event: MouseEvent): void {
    tablePos.set({
      x: initPos.x + event.clientX * $zoomStore - initMouseDown.x,
      y: initPos.y + event.clientY * $zoomStore - initMouseDown.y,
    });
  }

  const toHighlight = store.table.getTableToHighlight(name);
  const fromHighlight = store.table.getTableFromHighlight(name);

  const columnEnter = (column: ColumnFkSchema) => {
    store.table.setTableFromHighlight(table?.name ?? "", column.name);
    store.table.setTableToHighlight(column.fk.table, column.fk.column);
  };
  const columnLeave = (column: ColumnFkSchema) => {
    store.table.setTableFromHighlight(table?.name ?? "", "");
    store.table.setTableToHighlight(column.fk.table, "");
  };

  const dblClickInternal = () => dblClick(getTableData());
  const clickInternal = () => click(getTableData());
  const contextMenuInternal = () => contextMenu(getTableData());
</script>

<g>
  <foreignObject
    x={$tablePos ? $tablePos.x - width / 2 : 0}
    y={$tablePos ? $tablePos.y - height / 2 : 0}
    {width}
    {height}
    data-testid="table-{name}"
  >
    <!-- <div bind:this={tableElem}
      class="table"
      class:disable-select={moveInProgress}
      on:mousedown={mouseDownInternal}
      on:dblclick={dblClickInternal}
      on:click={clickInternal}
      on:contextmenu={contextMenuInternal}>
      <div class="table-name" colspan="2">
        {name}
      </div>
      {#each table?.columns ?? [] as column}
        <Column
          {column}
          {columnEnter}
          {columnLeave}
          highlight={$toHighlight === column.name
            ? Highlight.To
            : $fromHighlight === column.name
            ? Highlight.From
            : Highlight.None}
        />
      {/each}
    </div> -->
    <table
      bind:this={tableElem}
      class:disable-select={moveInProgress}
      on:mousedown={mouseDownInternal}
      on:dblclick={dblClickInternal}
      on:click={clickInternal}
      on:contextmenu={contextMenuInternal}
    >
      <thead>
        <tr><th colspan="4">{name}</th></tr>
      </thead>
      <tbody>
        {#each table?.columns ?? [] as column}
          <Column
            {column}
            {columnEnter}
            {columnLeave}
            highlight={$toHighlight === column.name
              ? Highlight.To
              : $fromHighlight === column.name
              ? Highlight.From
              : Highlight.None}
          />
        {/each}
      </tbody>
    </table>
  </foreignObject>
</g>
