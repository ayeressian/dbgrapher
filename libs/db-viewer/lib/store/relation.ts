import { isColumnFk } from "../schema";
import { derived, get, type Readable } from "svelte/store";
import type { Store } from "./store";

type Relation = {
  key: string;
  fromTable: string;
  toTable: string;
  fromColumn: string;
  toColumn: string;
  oneTo: boolean;
  toMany: boolean;
};

export class RelationStore {
  relations?: Readable<Relation[]>;

  constructor(private store: Store) {
    this.relations = derived(store.schema.schema, ($schema) => {
      const relations: Relation[] = [];
      let index = 0;
      $schema.tables.forEach((table) => {
        table.columns.forEach((column) => {
          if (isColumnFk(column)) {
            relations.push({
              key: `${table.name}_${column.name}_${column.fk.table}_${column.fk.column}`,
              toTable: table.name,
              fromTable: column.fk.table,
              toColumn: column.name,
              fromColumn: column.fk.column,
              oneTo: !!column.nn,
              toMany: !column.uq,
            });
            ++index;
          }
        });
      });
      return relations;
    });
  }

  getRelations(fromTable: string, toTable: string) {
    const realtionsVal = get(this.relations!);
    const r = realtionsVal.filter(
      (relation) =>
        relation.fromTable === fromTable && relation.toTable === toTable
    );
    return r;
  }
}
