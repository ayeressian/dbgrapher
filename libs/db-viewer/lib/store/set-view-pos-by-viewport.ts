import type Point from "../point";
import type { RectSize } from "./schema";
import type { Viewport } from "./view";

type Args = {
  viewport: Viewport;
  viewSizeAfterZoom: RectSize;
  tablePoses: Point[];
  currentPos: Point;
  viewBoundSize: RectSize;
};

const centerByTablesWeight = ({
  tablePoses,
  viewBoundSize,
}: Pick<Args, "tablePoses" | "viewBoundSize">): Point => {
  const tablePosesLength = [...tablePoses.values()].length;
  const totalSum = tablePoses.reduce(
    (acc, tablePos) => {
      acc.xTotal += tablePos.x;
      acc.yTotal += tablePos.y;
      return acc;
    },
    { xTotal: 0, yTotal: 0 }
  );

  return {
    x: totalSum.xTotal / tablePosesLength - viewBoundSize.width / 2,
    y: totalSum.yTotal / tablePosesLength - viewBoundSize.height / 2,
  };
};

const centerByTables = ({
  tablePoses,
  viewSizeAfterZoom,
}: Pick<Args, "tablePoses" | "viewSizeAfterZoom">): Point => {
  const posXes = tablePoses.map((item) => item.x);
  const posYes = tablePoses.map((item) => item.y);
  const minXPos = Math.min(...posXes);
  const maxXPos = Math.max(...posXes);
  const minYPos = Math.min(...posYes);
  const maxYPos = Math.max(...posYes);

  return {
    x: (minXPos + maxXPos) / 2 - viewSizeAfterZoom.width / 2,
    y: (minYPos + maxYPos) / 2 - viewSizeAfterZoom.height / 2,
  };
};

export const setViewPosByViewport = (args: Args): Point => {
  switch (args.viewport) {
    case "centerByTablesWeight": 
      return centerByTablesWeight(args);
    case "centerByTables":
      return centerByTables(args);
    case "center":
      return {
        x: 0,
        y: 0,
      };
    case "noChange":
      return args.currentPos;
  }
};
