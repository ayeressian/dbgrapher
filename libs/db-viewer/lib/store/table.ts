import type Point from "../point";
import { get, type Writable } from "svelte/store";
import type { RectSize } from "./schema";
import type { TableSchema } from "lib/schema";

type TableSize = RectSize;

export class TableStore {
  tableSizes = new Map<string, Writable<TableSize>>();
  tablePoses = new Map<string, Writable<Point>>();
  tableFromHighlight = new Map<string, Writable<string>>();
  tableToHighlight = new Map<string, Writable<string>>();

  getTableSize(tableName: string) {
    return this.tableSizes.get(tableName);
  }
  getTablePos(tableName: string) {
    return this.tablePoses.get(tableName);
  }
  getTableFromHighlight(tableName: string) {
    return this.tableFromHighlight.get(tableName);
  }
  getTableToHighlight(tableName: string) {
    return this.tableToHighlight.get(tableName);
  }
  setTableFromHighlight(tableName: string, column: string) {
    this.tableFromHighlight.get(tableName)?.set(column);
  }
  setTableToHighlight(tableName: string, column: string) {
    this.tableToHighlight.get(tableName)?.set(column);
  }

  updateTablePoses(tableSchemas: TableSchema[]) {
    tableSchemas.forEach((tableSchema) => {
      const pos = this.getTablePos(tableSchema.name)!;
      tableSchema.pos = get(pos);
    });
  }
}
