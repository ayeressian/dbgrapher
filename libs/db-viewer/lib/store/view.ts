import { derived, get, writable, type Writable } from "svelte/store";
import type Point from "../point";
import type { RectSize } from "./schema";
import { setViewPosByViewport } from "./set-view-pos-by-viewport";
import type { Store } from "./store";

const SCROLL_BUFFER = 0;

type ViewSize = RectSize;
type ViewPos = Point;
type ViewBound = { left: number; right: number; top: number; bottom: number };
export type Viewport =
  | "noChange"
  | "centerByTablesWeight"
  | "center"
  | "centerByTables";

const DEFAULT_ZOOM = 1;

export class ViewStore {
  viewSize = writable<ViewSize>({ width: 0, height: 0 });
  zoom = writable(DEFAULT_ZOOM);
  viewPos = writable<ViewPos>({ x: 0, y: 0 });
  viewBound = writable<ViewBound>();
  viewSizeAfterZoom = derived(
    [this.zoom, this.viewSize],
    ([zoomVal, viewSizeVal]) => {
      return {
        width: zoomVal * viewSizeVal.width,
        height: zoomVal * viewSizeVal.height,
      };
    }
  );
  viewBoundSize = derived(this.viewBound, (val) => {
    return {
      width: val.right - val.left,
      height: val.bottom - val.top,
    };
  });
  viewportStore = writable<Viewport>("center");
  viewportApply = writable(false);

  constructor(private store: Store) {
    this.viewSizeAfterZoom.subscribe(this.getUpdateViewBound());
    this.viewPos.subscribe(this.getUpdateViewBound());

    this.store.schema.schema.subscribe(this.applyViewport.bind(this));
    this.viewSizeAfterZoom.subscribe(this.applyViewport.bind(this));
  }

  private getTableStore() {
    return this.store.table;
  }

  getUpdateViewBound() {
    return this.updateViewBound.bind(this);
  }

  updateViewBound() {
    const viewPosVal = get(this.viewPos);
    const viewSizeAfterZoomVal = get(this.viewSizeAfterZoom);
    const newViewBound = {
      left: viewPosVal.x,
      right: viewPosVal.x + viewSizeAfterZoomVal.width,
      top: viewPosVal.y,
      bottom: viewPosVal.y + viewSizeAfterZoomVal.height,
    };
    for (const [tableName, pos] of this.getTableStore().tablePoses.entries()) {
      const tableSize = get(
        this.getTableStore().getTableSize(tableName) as Writable<RectSize>
      );
      const tablePosValue = get(pos);

      const tableWidth = tableSize?.width ?? 0;
      const tableHeight = tableSize?.height ?? 0;

      const right = tablePosValue.x + tableWidth / 2;
      if (right > newViewBound.right) {
        newViewBound.right = right;
      }

      const left = tablePosValue.x - tableWidth / 2;
      if (left < newViewBound.left) {
        newViewBound.left = left;
      }

      const top = tablePosValue.y - tableHeight / 2;
      if (top < newViewBound.top) {
        newViewBound.top = top;
      }

      const bottom = tablePosValue.y + tableHeight / 2;
      if (bottom > newViewBound.bottom) {
        newViewBound.bottom = bottom;
      }
    }

    newViewBound.left -= SCROLL_BUFFER;
    newViewBound.right += SCROLL_BUFFER;
    newViewBound.top -= SCROLL_BUFFER;
    newViewBound.bottom += SCROLL_BUFFER;

    this.viewBound.set(newViewBound);
  }

  applyViewport() {
    const viewportApplyVal = get(this.viewportApply);
    const schema = get(this.store.schema.schema);
    const viewSizeAfterZoomVal = get(this.viewSizeAfterZoom);

    if (
      schema.tables.length === 0 ||
      viewSizeAfterZoomVal.width === 0 ||
      viewportApplyVal
    )
      return;

    const tablePoseValues = [...this.store.table.tablePoses.values()].map(
      (tablePos) => get(tablePos)
    );
    this.viewPos.set(
      setViewPosByViewport({
        viewport: get(this.viewportStore),
        viewSizeAfterZoom: get(this.viewSizeAfterZoom),
        tablePoses: tablePoseValues,
        currentPos: get(this.viewPos),
        viewBoundSize: get(this.viewBoundSize),
      })
    );
    this.viewportApply.set(true);
  }
}
