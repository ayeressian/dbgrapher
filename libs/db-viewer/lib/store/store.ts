import { RelationStore } from "./relation";
import { SchemaStore } from "./schema";
import { TableStore } from "./table";
import { ViewStore } from "./view";

export class Store {
  schema = new SchemaStore(this);
  table = new TableStore();
  view = new ViewStore(this);
  relation = new RelationStore(this);
}
