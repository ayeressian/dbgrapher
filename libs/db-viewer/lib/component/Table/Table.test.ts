import schoolSchema from '../../../src/school';
import { setSchema } from '../../store/schema';
import { render, RenderResult, fireEvent, cleanup } from '@testing-library/svelte';
import Table from './Table.svelte';
import { beforeEach, describe, expect, it } from 'vitest';
import type Point from '../../point';

describe(Table.name, () => {
	let component: RenderResult;

	beforeEach(async () => {
		setSchema(schoolSchema);
		cleanup();
		component = render(Table, {
			name: 'school',
			zoom: 1,
		});
	});
	it('should render properly', () => {
		expect(component).toMatchSnapshot();
	});

	describe('when table is being moved', () => {
		const initCord = {
			x: 200,
			y: 300
		};

		const moveCord = {
			x: 500,
			y: 600
		};

		let schoolSchemaPos: Point;

		beforeEach(async () => {
			schoolSchemaPos = schoolSchema.tables.find((table) => table.name === 'school')!.pos;
			const schoolGElem = component.getByTestId('table-school').parentElement!;
			const mouseDown = new MouseEvent('mousedown', {
				clientX: initCord.x,
				clientY: initCord.y
			});
			await fireEvent(schoolGElem, mouseDown);
			const mouseMove = new MouseEvent('mousemove', {
				clientX: moveCord.x,
				clientY: moveCord.y
			});
			await fireEvent(document, mouseMove);
			const mouseUp = new MouseEvent('mouseup');
			await fireEvent(document, mouseUp);
		});

		it('should have correct cordinates', () => {
			const schoolElem = component.getByTestId('table-school');
			expect(schoolElem.getAttribute('x')).toEqual(String(schoolSchemaPos.x + moveCord.x - initCord.x));
			expect(schoolElem.getAttribute('y')).toEqual(String(schoolSchemaPos.y + moveCord.y - initCord.y));
		});
	});
});
