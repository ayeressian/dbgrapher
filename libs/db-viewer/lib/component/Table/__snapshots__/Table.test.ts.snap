// Vitest Snapshot v1

exports[`Table$ > should render properly 1`] = `
{
  "component": Table$ {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {},
      "ctx": [
        "school",
        0,
        0,
        <table>
          <thead>
            <tr>
              <th
                colspan="3"
              >
                school
              </th>
            </tr>
          </thead>
           
          <tbody>
            <tr>
              <td />
               
              <td>
                id
              </td>
               
              <td>
                int
              </td>
            </tr>
            <tr>
              <td />
               
              <td>
                cpacity
              </td>
               
              <td>
                int
              </td>
            </tr>
            <tr>
              <td />
               
              <td>
                name
              </td>
               
              <td>
                string
              </td>
            </tr>
            <tr>
              <td />
               
              <td>
                address
              </td>
               
              <td>
                string
              </td>
            </tr>
          </tbody>
        </table>,
        false,
        {
          "x": 722,
          "y": 560,
        },
        {
          "columns": [
            {
              "name": "id",
              "pk": true,
              "type": "int",
            },
            {
              "name": "cpacity",
              "type": "int",
            },
            {
              "name": "name",
              "type": "string",
            },
            {
              "name": "address",
              "type": "string",
            },
          ],
          "name": "school",
          "pos": {
            "x": 722,
            "y": 560,
          },
        },
        {
          "set": [Function],
          "subscribe": [Function],
          "update": [Function],
        },
        [Function],
        1,
        [Function],
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
        [Function],
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {
        "name": 0,
        "zoom": 9,
      },
      "root": <div>
        <g>
          <foreignobject
            data-testid="table-school"
            height="0"
            width="0"
            x="722"
            y="560"
          >
            <table>
              <thead>
                <tr>
                  <th
                    colspan="3"
                  >
                    school
                  </th>
                </tr>
              </thead>
               
              <tbody>
                <tr>
                  <td />
                   
                  <td>
                    id
                  </td>
                   
                  <td>
                    int
                  </td>
                </tr>
                <tr>
                  <td />
                   
                  <td>
                    cpacity
                  </td>
                   
                  <td>
                    int
                  </td>
                </tr>
                <tr>
                  <td />
                   
                  <td>
                    name
                  </td>
                   
                  <td>
                    string
                  </td>
                </tr>
                <tr>
                  <td />
                   
                  <td>
                    address
                  </td>
                   
                  <td>
                    string
                  </td>
                </tr>
              </tbody>
            </table>
          </foreignobject>
        </g>
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <g>
        <foreignobject
          data-testid="table-school"
          height="0"
          width="0"
          x="722"
          y="560"
        >
          <table>
            <thead>
              <tr>
                <th
                  colspan="3"
                >
                  school
                </th>
              </tr>
            </thead>
             
            <tbody>
              <tr>
                <td />
                 
                <td>
                  id
                </td>
                 
                <td>
                  int
                </td>
              </tr>
              <tr>
                <td />
                 
                <td>
                  cpacity
                </td>
                 
                <td>
                  int
                </td>
              </tr>
              <tr>
                <td />
                 
                <td>
                  name
                </td>
                 
                <td>
                  string
                </td>
              </tr>
              <tr>
                <td />
                 
                <td>
                  address
                </td>
                 
                <td>
                  string
                </td>
              </tr>
            </tbody>
          </table>
        </foreignobject>
      </g>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
